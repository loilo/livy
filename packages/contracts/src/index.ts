export * from './closable-handler-interface.js'
export * from './formattable-handler-interface.js'
export * from './formatter-interface.js'
export * from './handler-interface.js'
export * from './log-level.js'
export * from './log-record.js'
export * from './logger-interface.js'
export * from './processable-handler-interface.js'
export * from './processor-interface.js'
export * from './resettable-interface.js'
