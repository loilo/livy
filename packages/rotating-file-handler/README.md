# `@livy/rotating-file-handler`

This [Livy](../../README.md#readme) handler stores log records to files that are rotated by date/time or file size and only a limited number of files is kept.

---

[**Synchronous logger support:**](../../README.md#synchronous-and-asynchronous-logging) yes

**Runtime:** Node.js

---

## Basic Example

```js
import { RotatingFileHandler } from '@livy/rotating-file-handler'

const handler = new RotatingFileHandler('logs-%date%.txt')
```

## Installation

Install it via npm:

```bash
npm install @livy/rotating-file-handler
```

## Options

The first argument to this handler's constructor is a file path to write to. It needs to contain a token based on the configured [`strategy`](#strategy).

An object of options can be passed to the constructor as the second argument.

The following options are available:

### `bubble`

**Type:** `boolean`

**Default:** `true`

**Description:** Controls whether records handled by this handler should bubble up to other handlers.

> See also: [Bubbling](../../README.md#bubbling)

### `formatter`

**Type:** [`FormatterInterface`](../contracts/README.md#formatterinterface)

**Default:** [`new LineFormatter()`](../line-formatter/README.md#readme)

**Description:** The formatter to use.

### `level`

**Type:** [`LogLevel`](../contracts/README.md#loglevel)

**Default:** `'debug'`

**Description:** Controls which log records should be handled based on their [log level](../../README.md#log-levels).

### `maxFiles`

**Type:** `number`

**Default:** `Infinity`

**Description:** Number of log files to keep at maximum.

### `strategy`

**Type:** `'max-age' | 'max-size'`

**Default:** `'max-age'`

**Description:** The strategy name:

- `max-age`: Separates log files by the date/time attached to log records thus creates a new file every year/month/day/hour/minute. It requires a `%date%` token in the file name.
- `max-size`: Separates log files by the size of the latest log file and thus creates a new file whenever the current file exceeds the size. It requires a `%appendix%` token in the file name.

### `threshold`

The value this option takes varies based on the [`strategy`](#strategy) option:

#### Strategy `max-age`

**Type:** `'minute' | 'hour' | 'day' | 'month' | 'year'`

**Default:** `'day'`

**Description:** The duration unit that separates individual log files. Using `'day'` rotates the log files daily, using `'hour'` rotates them every hour etc.

#### Strategy `max-size`

**Type:** `number | string`

**Description:** The file size that separates individual log files. May be a plain number of bytes or a file size string like `'200 kb'`, `'1.5M'` etc.

## Public API

### `bubble`

Controls whether records handled by this handler should bubble up to other handlers. Initially set through the [`bubble`](#bubble) option.

> See also: [Bubbling](../../README.md#bubbling)

### `close()`

This handler implements the [`ClosableHandlerInterface`](../contracts/README.md#closablehandlerinterface). On cleanup, it closes the internal file handler.

> You usually don't want to call this method manually. It is done automatically when a Node.js process exits / a browser page is closed.

### `defaultFormatter` (read-only)

The formatter used by this handler if no [`formatter`](#formatter) option is set.

### `formatter`

This handler supports formatters by implementing the [`FormattableHandlerInterface`](../contracts/README.md#formattablehandlerinterface).

### `level`

The minimum [log level](../../README.md#log-levels) of a log record to be considered by this handler. Initially set through the [`level`](#level) option.

### `processors`

This handler supports [processors](../../README.md#processors) by implementing the [`ProcessableHandlerInterface`](../contracts/README.md#processablehandlerinterface).

### `reset()`

This handler implements the [`ResettableInterface`](../contracts/README.md#resettableinterface). Resetting it resets all attached processors.

> You usually don't want to call this method manually on an individual handler. Consider calling it [on the logger](../logger/README.md#reset) instead.
