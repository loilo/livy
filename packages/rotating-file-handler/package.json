{
  "name": "@livy/rotating-file-handler",
  "type": "module",
  "version": "1.0.5",
  "description": "Stores Livy log records in rotating files",
  "keywords": [
    "logging",
    "logger",
    "logs",
    "log",
    "loglevel",
    "tools",
    "monolog",
    "handler",
    "size",
    "date",
    "filesize",
    "rotate"
  ],
  "author": "loilo <florian.reuschel@gmx.de>",
  "homepage": "https://github.com/loilo/livy/tree/master/packages/rotating-file-handler#readme",
  "license": "MIT",
  "exports": {
    "types": "./lib/rotating-file-handler.d.ts",
    "default": "./lib/rotating-file-handler.js"
  },
  "sideEffects": false,
  "directories": {
    "lib": "lib",
    "test": "test"
  },
  "files": [
    "lib"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": "https://github.com/loilo/livy/tree/master/packages/rotating-file-handler",
  "scripts": {
    "build": "del lib && tsc --declaration",
    "lint:readme": "remark --quiet --no-stdout -u validate-links README.md",
    "lint:source": "eslint \"src/**/*.ts\"",
    "test": "vitest run --coverage"
  },
  "bugs": {
    "url": "https://github.com/loilo/livy/issues"
  },
  "dependencies": {
    "@livy/contracts": "^1.1.2",
    "@livy/file-handler": "^1.0.4",
    "@livy/util": "^1.0.6",
    "@types/filesize-parser": "^1.5.0",
    "@types/luxon": "^3.2.0",
    "@types/node": "^18.11.18",
    "filesize-parser": "^1.5.0",
    "luxon": "^3.2.0"
  },
  "devDependencies": {
    "typescript": "^5.1.0",
    "@livy/bin": "file:../../bin",
    "@livy/test-utils": "file:../../test-utils"
  },
  "gitHead": "bd96a3e4d1e5870873b0e0b9e538486aea6343a2"
}
