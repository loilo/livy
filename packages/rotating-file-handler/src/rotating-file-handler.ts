import { isAbsolute, join, parse } from 'node:path'
import type {
  ClosableHandlerInterface,
  FormatterInterface,
  LogRecord,
} from '@livy/contracts'
import { FileHandler } from '@livy/file-handler'
import { AbstractSyncFormattingProcessingHandler } from '@livy/util/handlers/abstract-formatting-processing-handler'
import { AbstractLevelBubbleHandlerOptions } from '@livy/util/handlers/abstract-level-bubble-handler'
import { FormattableHandlerMixin } from '@livy/util/handlers/formattable-handler-mixin'
import {
  DurationUnit,
  MaxAgeStrategy,
  MaxAgeStrategyOptions,
} from './max-age-strategy.js'
import { MaxSizeStrategy, MaxSizeStrategyOptions } from './max-size-strategy.js'
import { RotationStrategyInterface } from './rotation-strategy.js'

interface BaseRotatingFileHandlerOptions
  extends AbstractLevelBubbleHandlerOptions {
  /**
   * The formatter to use
   */
  formatter: FormatterInterface

  /**
   * Number of log files to keep at maximum
   */
  maxFiles: number
}

export type RotatingFileHandlerOptions = BaseRotatingFileHandlerOptions &
  (MaxAgeStrategyOptions | MaxSizeStrategyOptions)

/**
 * Stores log records to files that are rotated by datetime or file size
 * and only a limited number of files is kept.
 */
export class RotatingFileHandler
  extends FormattableHandlerMixin(AbstractSyncFormattingProcessingHandler)
  implements ClosableHandlerInterface
{
  private fileHandler!: FileHandler
  private directory: string
  private maxFiles: number
  private rotationStrategyHandler: RotationStrategyInterface

  public constructor(
    pathTemplate: string,
    {
      maxFiles = Number.POSITIVE_INFINITY,
      strategy = 'max-age',
      threshold = 'day',
      formatter,
      ...options
    }: Partial<RotatingFileHandlerOptions> = {},
  ) {
    super(options)

    this.maxFiles = maxFiles
    this.explicitFormatter = formatter

    if (!isAbsolute(pathTemplate)) {
      pathTemplate = join(process.cwd(), pathTemplate)
    }

    const pathData = parse(pathTemplate)

    switch (strategy) {
      case 'max-age':
        this.rotationStrategyHandler = new MaxAgeStrategy(
          pathData.dir,
          pathData.base,
          threshold as DurationUnit,
        )
        break

      case 'max-size':
        this.rotationStrategyHandler = new MaxSizeStrategy(
          pathData.dir,
          pathData.base,
          threshold,
        )
        break

      default:
        throw new Error(`Invalid rotation strategy "${strategy}"`)
    }

    this.directory = pathData.dir
    this.updateFileHandler()
  }

  /**
   * Update the file handler to use the current filename
   */
  private updateFileHandler() {
    if (this.fileHandler !== undefined) {
      this.fileHandler.close()
    }

    this.fileHandler = new FileHandler(
      join(this.directory, this.rotationStrategyHandler.getCurrentFilename()),
      { level: this.level, formatter: this.formatter },
    )
  }

  /**
   * @inheritdoc
   */
  public close() {
    this.rotateIfNeeded()
  }

  /**
   * Rotate if needed
   */
  private rotateIfNeeded() {
    if (this.rotationStrategyHandler.shouldRotate()) {
      this.rotationStrategyHandler.rotate(this.maxFiles)
      this.updateFileHandler()
    }
  }

  /**
   * @inheritdoc
   */
  protected async write(record: LogRecord) {
    this.rotateIfNeeded()
    await this.fileHandler.handle(record)
  }

  /**
   * @inheritdoc
   */
  protected writeSync(record: LogRecord) {
    this.rotateIfNeeded()
    this.fileHandler.handleSync(record)
  }
}
