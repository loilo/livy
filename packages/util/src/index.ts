export type * from './formatters/abstract-batch-formatter.js'
export type * from './formatters/included-record-properties.js'
export type * from './formatters/line-formatter.js'
export type * from './handlers/abstract-batch-handler.js'
export type * from './handlers/abstract-formatting-processing-handler.js'
export type * from './handlers/abstract-level-bubble-handler.js'
export type * from './handlers/formattable-handler-mixin.js'
export type * from './handlers/is-closable-handler-interface.js'
export type * from './handlers/is-formattable-handler-interface.js'
export type * from './handlers/is-processable-handler-interface.js'
export type * from './handlers/is-sync-handler-interface.js'
export type * from './handlers/mirror-sync-handler-mixin.js'
export type * from './handlers/processable-handler-mixin.js'
export type * from './handlers/respect-level-mixin.js'
export type * from './console-adapter-interface.js'
export type * from './console-adapter.js'
export type * from './environment.js'
export type * from './gated-set.js'
export type * from './helpers.js'
export type * from './html.js'
export type * from './is-resettable-interface.js'
export type * from './mixin.js'
export type * from './timer.js'
export type * from './types.js'
export type * from './validatable-set.js'
