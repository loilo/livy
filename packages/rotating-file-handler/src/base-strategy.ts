import { existsSync, readdirSync, statSync, unlinkSync } from 'node:fs'
import { join } from 'node:path'
import { MaxAgeStrategyOptions } from './max-age-strategy.js'
import { MaxSizeStrategyOptions } from './max-size-strategy.js'
import { RotationStrategyInterface } from './rotation-strategy.js'

/**
 * Common functionality for all `RotatingFileHandler` strategies
 */
export abstract class BaseStrategy<
  T extends MaxAgeStrategyOptions | MaxSizeStrategyOptions,
> implements RotationStrategyInterface
{
  constructor(
    protected directory: string,
    protected filenameTemplate: string,
    protected threshold: T['threshold'],
  ) {
    if (!existsSync(directory)) {
      throw new Error(
        `Directory for rotating log files "${directory}" does not exist`,
      )
    }
  }

  /**
   * @inheritdoc
   */
  public abstract getCurrentFilename(): string

  /**
   * @inheritdoc
   */
  public abstract shouldRotate(): boolean

  /**
   * @inheritdoc
   */
  public abstract rotate(maxFiles: number): void

  /**
   * Compare (as in a custom sort callback) two filenames so that the older one goes first and the more recent one goes after
   * This is used to sort existing log files and delete the older ones
   *
   * @param a
   * @param b
   */
  protected abstract compareFilenames(a: string, b: string): number

  /**
   * Get a regular expression that matches log files generated by this strategy
   */
  protected abstract getFilenameRegex(): RegExp

  /**
   * Get a list of existing log files, sorted by recency
   */
  protected getExistingFiles() {
    const regex = this.getFilenameRegex()

    return readdirSync(this.directory)
      .filter(
        file => regex.test(file) && statSync(join(this.directory, file)).isFile,
      )
      .sort(this.compareFilenames.bind(this))
  }

  /**
   * Delete files that are no longer needed according to `maxFiles`
   *
   * @param maxFiles The number of log files to keep
   */
  protected deleteSurplusFiles(maxFiles: number) {
    const files = this.getExistingFiles()

    // If the current file name is *not* among the existing ones, we decrement
    // the number of allowed files since probably there will be
    // a new file created with the next logging entry
    if (!files.includes(this.getCurrentFilename())) {
      maxFiles--
    }

    let filesToDelete: string[]
    if (maxFiles === 0) {
      filesToDelete = files
    } else {
      filesToDelete = files.slice(0, -maxFiles)
    }

    for (const file of filesToDelete) {
      unlinkSync(join(this.directory, file))
    }
  }
}
