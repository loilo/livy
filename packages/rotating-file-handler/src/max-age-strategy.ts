import { existsSync } from 'node:fs'
import { replaceTokens, sanitizeRegex } from '@livy/util/helpers'
import { DateTime } from 'luxon'
import { BaseStrategy } from './base-strategy.js'
import { RotationStrategyInterface } from './rotation-strategy.js'

export interface MaxAgeStrategyOptions {
  /**
   * The strategy name
   */
  strategy: 'max-age'

  /**
   * The duration unit that separates individual log files
   */
  threshold: DurationUnit
}

export type DurationUnit = 'minute' | 'hour' | 'day' | 'month' | 'year'
type DateFormatter = (datetime: DateTime) => string

/**
 * Separates log files by the datetime attached to log records thus creates
 * a new file every year/month/day/hour/minute
 */
export class MaxAgeStrategy
  extends BaseStrategy<MaxAgeStrategyOptions>
  implements RotationStrategyInterface
{
  private cachedFilename?: string
  private durationUnit: DurationUnit
  private dateFormatter: DateFormatter

  constructor(
    directory: string,
    filenameTemplate: string,
    threshold: DurationUnit,
  ) {
    if (!filenameTemplate.includes('%date%')) {
      throw new Error(
        `Invalid filename template "${filenameTemplate}", must contain the %date% token.`,
      )
    }

    super(directory, filenameTemplate, threshold)

    this.durationUnit = threshold
    this.dateFormatter = this.getDateFormatter(threshold)
  }

  /**
   * Get a regular expression matching log file names
   */
  protected getFilenameRegex() {
    return new RegExp(
      `^${replaceTokens(sanitizeRegex(this.filenameTemplate), {
        date: '.+?',
      })}$`,
    )
  }

  /**
   * Get the date format for a unit
   *
   * @param unit The unit to get the date format for
   */
  private getDateFormatter(unit: DurationUnit): DateFormatter {
    return {
      minute: (datetime: DateTime) => datetime.toFormat('yyyy-MM-dd_HH-mm'),
      hour: (datetime: DateTime) => datetime.toFormat('yyyy-MM-dd_HH'),
      day: (datetime: DateTime) => datetime.toFormat('yyyy-MM-dd'),
      month: (datetime: DateTime) => datetime.toFormat('yyyy-MM'),
      year: (datetime: DateTime) => datetime.toFormat('yyyy'),
    }[unit]
  }

  /**
   * Get the name of the current file
   */
  public getCurrentFilename() {
    const date = this.getStartOfDurationUnit()

    return replaceTokens(this.filenameTemplate, {
      date: this.dateFormatter(date),
    })
  }

  /**
   * Get the current date set to a unit-relative starting point
   */
  protected getStartOfDurationUnit() {
    return DateTime.local().startOf(this.durationUnit)
  }

  /**
   * @inheritdoc
   */
  public shouldRotate() {
    const filename = this.getCurrentFilename()

    if (this.cachedFilename === filename) {
      return false
    }

    this.cachedFilename = filename

    // Non-existing file indicates that there may be pre-existing files to rotate
    return !existsSync(filename)
  }

  /**
   * @inheritdoc
   */
  protected compareFilenames(a: string, b: string) {
    return a.localeCompare(b)
  }

  /**
   * @inheritdoc
   */
  public rotate(maxFiles: number) {
    this.deleteSurplusFiles(maxFiles)
  }
}
