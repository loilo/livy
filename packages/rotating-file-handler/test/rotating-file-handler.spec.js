import * as fs from 'node:fs'
import { describe, expect, it, afterEach, vi } from 'vitest'
import { DateTime } from 'luxon'
import { RotatingFileHandler } from '../src/rotating-file-handler'

vi.mock(
  'luxon',
  livyTestGlobals.getMockedModule(
    import('@livy/test-utils/mocks/modules/luxon.js'),
  ),
)

const { record, TEST_CONSTANTS, date, MockFormatter } = livyTestGlobals

// Fixate timezone for file names to the test constant
vi.mock('../src/max-age-strategy', async importOriginal => {
  const { MaxAgeStrategy } = await importOriginal()

  return {
    MaxAgeStrategy: class extends MaxAgeStrategy {
      getStartOfDurationUnit() {
        return super.getStartOfDurationUnit().setZone(TEST_CONSTANTS.TIMEZONE)
      }
    },
  }
})

/**
 * Write a number of records to a handler at a fixated date
 *
 * @param {HandlerInterface} handler The handler to write to
 * @param {'sync'|'async'}   mode    Whether to use `handle` and continue or `handleAsync` and await
 * @param {string}           isoDate The fixated date as an ISO 8601 string
 * @param {string[]}         levels  An array of log levels that will be mapped to invoked records
 */
async function writeRecords(handler, mode, isoDate, levels) {
  const createRecord = record.with({
    datetime: DateTime.fromISO(isoDate).setZone(TEST_CONSTANTS.TIMEZONE),
    message: 'Test RotatingFileHandler',
  })
  const records = levels.map(level => createRecord({ level }))

  DateTime.mock.fixate(isoDate)
  for (const record of records) {
    if (mode === 'sync') {
      handler.handleSync(record)
    } else {
      await handler.handle(record)
    }
  }
  DateTime.mock.release()
}

describe('@livy/rotating-file-handler', () => {
  afterEach(() => {
    date.release()
    fs.__reset()
  })

  it('should use daily rotation with default settings', () => {
    const handler = new RotatingFileHandler('/logfile-%date%.txt')

    const handle = writeRecords.bind(null, handler, 'sync')
    handle('2015-08-15T14:30:00+02:00', ['debug', 'info'])
    handle('2015-08-16T05:12:21+02:00', ['notice', 'warning'])
    handle('2015-08-16T13:13:13+02:00', ['error', 'critical'])
    handle('2015-08-17T13:13:13+02:00', ['alert', 'emergency'])

    expect(fs.toJSON()).toMatchSnapshot()
  })

  it('should use daily rotation with default settings (async)', async () => {
    const handler = new RotatingFileHandler('/logfile-%date%.txt')

    const handle = writeRecords.bind(null, handler, 'async')
    await handle('2015-08-15T14:30:00+02:00', ['debug', 'info'])
    await handle('2015-08-16T05:12:21+02:00', ['notice', 'warning'])
    await handle('2015-08-16T13:13:13+02:00', ['error', 'critical'])
    await handle('2015-08-17T13:13:13+02:00', ['alert', 'emergency'])

    expect(fs.toJSON()).toMatchSnapshot()
  })

  it('should rotate on handler close', () => {
    const handler = new RotatingFileHandler('/logfile-%date%.txt', {
      maxFiles: 1,
    })

    writeRecords(handler, 'sync', '2015-08-15T14:30:00+02:00', ['debug'])

    date.fixate('2015-08-16T05:12:21+02:00')

    handler.close()

    const files = fs.toJSON()
    expect(typeof files).toBe('object')
    expect(Object.keys(files)).toHaveLength(0)
  })

  it('should perform max-age rotation minutely', () => {
    const handler = new RotatingFileHandler('/logfile-%date%.txt', {
      threshold: 'minute',
    })

    const handle = writeRecords.bind(null, handler, 'sync')

    handle('2015-08-15T14:30:00+02:00', ['debug', 'info'])
    handle('2015-08-16T05:12:21+02:00', ['notice', 'warning'])
    handle('2015-08-16T05:12:59+02:00', ['error', 'critical'])
    handle('2015-08-16T13:13:13+02:00', ['alert', 'emergency'])

    expect(fs.toJSON()).toMatchSnapshot()
  })

  it('should perform max-age rotation hourly', () => {
    const handler = new RotatingFileHandler('/logfile-%date%.txt', {
      threshold: 'hour',
    })

    const handle = writeRecords.bind(null, handler, 'sync')

    handle('2015-08-15T14:30:00+02:00', ['debug', 'info'])
    handle('2015-08-16T05:12:21+02:00', ['notice', 'warning'])
    handle('2015-08-16T05:12:59+02:00', ['error', 'critical'])
    handle('2015-08-16T13:13:13+02:00', ['alert', 'emergency'])

    expect(fs.toJSON()).toMatchSnapshot()
  })

  it('should perform max-age rotation daily', () => {
    const handler = new RotatingFileHandler('/logfile-%date%.txt', {
      threshold: 'day',
    })

    const handle = writeRecords.bind(null, handler, 'sync')

    handle('2015-08-15T14:30:00+02:00', ['debug', 'info'])
    handle('2015-08-16T05:12:21+02:00', ['notice', 'warning'])
    handle('2015-08-16T05:12:59+02:00', ['error', 'critical'])
    handle('2015-08-16T13:13:13+02:00', ['alert', 'emergency'])

    expect(fs.toJSON()).toMatchSnapshot()
  })

  it('should perform max-age rotation monthly', () => {
    const handler = new RotatingFileHandler('/logfile-%date%.txt', {
      threshold: 'month',
    })

    const handle = writeRecords.bind(null, handler, 'sync')

    handle('2015-08-15T14:30:00+02:00', ['debug', 'info'])
    handle('2015-08-16T05:12:21+02:00', ['notice', 'warning'])
    handle('2015-09-02T05:12:59+02:00', ['error', 'critical'])
    handle('2016-01-16T13:13:13+02:00', ['alert', 'emergency'])

    expect(fs.toJSON()).toMatchSnapshot()
  })

  it('should perform max-age rotation yearly', () => {
    const handler = new RotatingFileHandler('/logfile-%date%.txt', {
      threshold: 'year',
    })

    const handle = writeRecords.bind(null, handler, 'sync')

    handle('2015-08-15T14:30:00+02:00', ['debug', 'info'])
    handle('2015-08-16T05:12:21+02:00', ['notice', 'warning'])
    handle('2016-09-02T05:12:59+02:00', ['error', 'critical'])
    handle('2017-01-16T13:13:13+02:00', ['alert', 'emergency'])

    expect(fs.toJSON()).toMatchSnapshot()
  })

  it('should do max-size rotation', () => {
    const handler = new RotatingFileHandler('/logfile%appendix%.txt', {
      strategy: 'max-size',
      threshold: '100 Bytes',
    })

    handler.handleSync(record('debug', 'Test RotatingFileHandler'))
    handler.handleSync(record('info', 'Test RotatingFileHandler'))
    handler.handleSync(record('notice', 'Test RotatingFileHandler'))
    handler.handleSync(record('warning', 'Test RotatingFileHandler'))
    handler.handleSync(record('error', 'Test RotatingFileHandler'))

    expect(fs.toJSON()).toMatchSnapshot()
  })

  it('should not accept a non-existent directory', () => {
    expect(() => {
      new RotatingFileHandler('/invalid/logfile-%date%.txt')
    }).toThrowError(
      new Error('Directory for rotating log files "/invalid" does not exist'),
    )
  })

  it('should not accept an invalid rotation strategy', () => {
    expect(() => {
      new RotatingFileHandler('/logfile.txt', {
        strategy: 'invalid',
      })
    }).toThrowError(new Error('Invalid rotation strategy "invalid"'))
  })

  it('should require a %date% token in max-age rotation', () => {
    expect(() => {
      new RotatingFileHandler('/logfile.txt', {
        strategy: 'max-age',
      })
    }).toThrowError(
      new Error(
        'Invalid filename template "logfile.txt", must contain the %date% token.',
      ),
    )
  })

  it('should require a %appendix% token in max-size rotation', () => {
    expect(() => {
      new RotatingFileHandler('/logfile.txt', {
        strategy: 'max-size',
      })
    }).toThrowError(
      new Error(
        'Invalid filename template "logfile.txt", must contain the %appendix% token.',
      ),
    )
  })

  it('should ignore a %date% token in the directory', () => {
    fs.mkdirSync('/%date%')

    expect(() => {
      new RotatingFileHandler('/%date%/logfile.txt', {
        strategy: 'max-age',
      })
    }).toThrowError(
      new Error(
        'Invalid filename template "logfile.txt", must contain the %date% token.',
      ),
    )
  })

  it('should ignore an %appendix% token in the directory', () => {
    fs.mkdirSync('/%appendix%')

    expect(() => {
      new RotatingFileHandler('/logfile.txt', {
        strategy: 'max-size',
      })
    }).toThrowError(
      new Error(
        'Invalid filename template "logfile.txt", must contain the %appendix% token.',
      ),
    )
  })

  it('should respect the "formatter" option', () => {
    const createRecord = record.with({
      level: 'info',
      message: '',
      datetime: DateTime.fromISO(TEST_CONSTANTS.DATE_ISO).setZone(
        TEST_CONSTANTS.TIMEZONE,
      ),
    })
    date.fixate(TEST_CONSTANTS.DATE_ISO)

    const formatter = new MockFormatter()

    const handler = new RotatingFileHandler('/logfile-%date%.txt', {
      formatter,
    })

    handler.handleSync(createRecord())

    expect(handler.formatter).toBe(formatter)
    expect(formatter.format).toHaveBeenCalledTimes(2)
    expect(formatter.format).toHaveBeenLastCalledWith(createRecord())
  })

  it('should respect the "level" option', () => {
    const createRecord = record.with({
      message: '',
      datetime: DateTime.fromISO(TEST_CONSTANTS.DATE_ISO).setZone(
        TEST_CONSTANTS.TIMEZONE,
      ),
    })
    date.fixate(TEST_CONSTANTS.DATE_ISO)

    const handler = new RotatingFileHandler('/logfile-%date%.txt', {
      level: 'notice',
    })

    expect(handler.isHandling('info')).toBe(false)
    expect(handler.isHandling('notice')).toBe(true)
    handler.handleSync(createRecord({ level: 'info' }))
    handler.handleSync(createRecord({ level: 'notice' }))
    expect(fs.toJSON()).toMatchSnapshot()
  })

  it('should respect the "bubble" option', () => {
    const bubblingHandler = new RotatingFileHandler('/logfile-%date%.txt')
    const nonBubblingHandler = new RotatingFileHandler('/logfile-%date%.txt', {
      bubble: false,
    })

    expect(bubblingHandler.handleSync(record('debug'))).toBe(false)
    expect(nonBubblingHandler.handleSync(record('debug'))).toBe(true)
  })
})
