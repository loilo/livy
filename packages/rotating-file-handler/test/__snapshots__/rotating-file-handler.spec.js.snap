// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@livy/rotating-file-handler > should do max-size rotation 1`] = `
{
  "/logfile.1.txt": "2015-08-15 14:30:00 NOTICE Test RotatingFileHandler {}
2015-08-15 14:30:00 WARNING Test RotatingFileHandler {}
",
  "/logfile.2.txt": "2015-08-15 14:30:00 DEBUG Test RotatingFileHandler {}
2015-08-15 14:30:00 INFO Test RotatingFileHandler {}
",
  "/logfile.txt": "2015-08-15 14:30:00 ERROR Test RotatingFileHandler {}
",
}
`;

exports[`@livy/rotating-file-handler > should perform max-age rotation daily 1`] = `
{
  "/logfile-2015-08-15.txt": "2015-08-15 14:30:00 DEBUG Test RotatingFileHandler {}
2015-08-15 14:30:00 INFO Test RotatingFileHandler {}
",
  "/logfile-2015-08-16.txt": "2015-08-16 05:12:21 NOTICE Test RotatingFileHandler {}
2015-08-16 05:12:21 WARNING Test RotatingFileHandler {}
2015-08-16 05:12:59 ERROR Test RotatingFileHandler {}
2015-08-16 05:12:59 CRITICAL Test RotatingFileHandler {}
2015-08-16 13:13:13 ALERT Test RotatingFileHandler {}
2015-08-16 13:13:13 EMERGENCY Test RotatingFileHandler {}
",
}
`;

exports[`@livy/rotating-file-handler > should perform max-age rotation hourly 1`] = `
{
  "/logfile-2015-08-15_14.txt": "2015-08-15 14:30:00 DEBUG Test RotatingFileHandler {}
2015-08-15 14:30:00 INFO Test RotatingFileHandler {}
",
  "/logfile-2015-08-16_05.txt": "2015-08-16 05:12:21 NOTICE Test RotatingFileHandler {}
2015-08-16 05:12:21 WARNING Test RotatingFileHandler {}
2015-08-16 05:12:59 ERROR Test RotatingFileHandler {}
2015-08-16 05:12:59 CRITICAL Test RotatingFileHandler {}
",
  "/logfile-2015-08-16_13.txt": "2015-08-16 13:13:13 ALERT Test RotatingFileHandler {}
2015-08-16 13:13:13 EMERGENCY Test RotatingFileHandler {}
",
}
`;

exports[`@livy/rotating-file-handler > should perform max-age rotation minutely 1`] = `
{
  "/logfile-2015-08-15_14-30.txt": "2015-08-15 14:30:00 DEBUG Test RotatingFileHandler {}
2015-08-15 14:30:00 INFO Test RotatingFileHandler {}
",
  "/logfile-2015-08-16_05-12.txt": "2015-08-16 05:12:21 NOTICE Test RotatingFileHandler {}
2015-08-16 05:12:21 WARNING Test RotatingFileHandler {}
2015-08-16 05:12:59 ERROR Test RotatingFileHandler {}
2015-08-16 05:12:59 CRITICAL Test RotatingFileHandler {}
",
  "/logfile-2015-08-16_13-13.txt": "2015-08-16 13:13:13 ALERT Test RotatingFileHandler {}
2015-08-16 13:13:13 EMERGENCY Test RotatingFileHandler {}
",
}
`;

exports[`@livy/rotating-file-handler > should perform max-age rotation monthly 1`] = `
{
  "/logfile-2015-08.txt": "2015-08-15 14:30:00 DEBUG Test RotatingFileHandler {}
2015-08-15 14:30:00 INFO Test RotatingFileHandler {}
2015-08-16 05:12:21 NOTICE Test RotatingFileHandler {}
2015-08-16 05:12:21 WARNING Test RotatingFileHandler {}
",
  "/logfile-2015-09.txt": "2015-09-02 05:12:59 ERROR Test RotatingFileHandler {}
2015-09-02 05:12:59 CRITICAL Test RotatingFileHandler {}
",
  "/logfile-2016-01.txt": "2016-01-16 12:13:13 ALERT Test RotatingFileHandler {}
2016-01-16 12:13:13 EMERGENCY Test RotatingFileHandler {}
",
}
`;

exports[`@livy/rotating-file-handler > should perform max-age rotation yearly 1`] = `
{
  "/logfile-2015.txt": "2015-08-15 14:30:00 DEBUG Test RotatingFileHandler {}
2015-08-15 14:30:00 INFO Test RotatingFileHandler {}
2015-08-16 05:12:21 NOTICE Test RotatingFileHandler {}
2015-08-16 05:12:21 WARNING Test RotatingFileHandler {}
",
  "/logfile-2016.txt": "2016-09-02 05:12:59 ERROR Test RotatingFileHandler {}
2016-09-02 05:12:59 CRITICAL Test RotatingFileHandler {}
",
  "/logfile-2017.txt": "2017-01-16 12:13:13 ALERT Test RotatingFileHandler {}
2017-01-16 12:13:13 EMERGENCY Test RotatingFileHandler {}
",
}
`;

exports[`@livy/rotating-file-handler > should respect the "level" option 1`] = `
{
  "/logfile-2015-08-15.txt": "2015-08-15 14:30:00 NOTICE  {}
",
}
`;

exports[`@livy/rotating-file-handler > should use daily rotation with default settings (async) 1`] = `
{
  "/logfile-2015-08-15.txt": "2015-08-15 14:30:00 DEBUG Test RotatingFileHandler {}
2015-08-15 14:30:00 INFO Test RotatingFileHandler {}
",
  "/logfile-2015-08-16.txt": "2015-08-16 05:12:21 NOTICE Test RotatingFileHandler {}
2015-08-16 05:12:21 WARNING Test RotatingFileHandler {}
2015-08-16 13:13:13 ERROR Test RotatingFileHandler {}
2015-08-16 13:13:13 CRITICAL Test RotatingFileHandler {}
",
  "/logfile-2015-08-17.txt": "2015-08-17 13:13:13 ALERT Test RotatingFileHandler {}
2015-08-17 13:13:13 EMERGENCY Test RotatingFileHandler {}
",
}
`;

exports[`@livy/rotating-file-handler > should use daily rotation with default settings 1`] = `
{
  "/logfile-2015-08-15.txt": "2015-08-15 14:30:00 DEBUG Test RotatingFileHandler {}
2015-08-15 14:30:00 INFO Test RotatingFileHandler {}
",
  "/logfile-2015-08-16.txt": "2015-08-16 05:12:21 NOTICE Test RotatingFileHandler {}
2015-08-16 05:12:21 WARNING Test RotatingFileHandler {}
2015-08-16 13:13:13 ERROR Test RotatingFileHandler {}
2015-08-16 13:13:13 CRITICAL Test RotatingFileHandler {}
",
  "/logfile-2015-08-17.txt": "2015-08-17 13:13:13 ALERT Test RotatingFileHandler {}
2015-08-17 13:13:13 EMERGENCY Test RotatingFileHandler {}
",
}
`;
